patches/server/0962-Registry-Modification-API.patch

```diff
@@ -0,0 +1,1239 @@
++import net.minecraft.core.Registry;
++import net.minecraft.resources.ResourceKey;
++import org.bukkit.Keyed;
++import org.checkerframework.checker.nullness.qual.Nullable;
++
++@Deprecated
++public class DelayedRegistryEntry<M, T extends Keyed> implements RegistryEntry<M, T> {
++
++    private final RegistryEntry<M, T> delegate;
++
++    public DelayedRegistryEntry(final RegistryEntry<M, T> delegate) {
++        this.delegate = delegate;
++    }
++
++    @Override
++    public ResourceKey<? extends Registry<M>> resourceKey() {
++        return this.delegate.resourceKey();
++    }
++
++    @Override
++    public RegistryKey2<T> key() {
++        return this.delegate.key();
++    }
++
++    @Override
++    public org.bukkit.Registry<T> createRegistry(final @Nullable Registry<M> registry) {
++        if (registry == null && !PaperRegistryAccess.instance().delayedValidCheck().getAsBoolean()) {
++            return new DelayedRegistry<>(() -> this.delegate.createRegistry(PaperRegistryAccess.instance().getDelayedRegistryAccess().registryOrThrow(this.resourceKey())), PaperRegistryAccess.instance().delayedValidCheck());
++        }
++        return this.delegate.createRegistry(registry);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/registry/legacy/package-info.java b/src/main/java/io/papermc/paper/registry/legacy/package-info.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..4396982af55872fafbfeaf8161ad6f392726c773
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/registry/legacy/package-info.java
+@@ -0,0 +1,5 @@
++@DefaultQualifier(NonNull.class)
++package io.papermc.paper.registry.legacy;
++
++import org.checkerframework.checker.nullness.qual.NonNull;
++import org.checkerframework.framework.qual.DefaultQualifier;
+diff --git a/src/main/java/io/papermc/paper/registry/package-info.java b/src/main/java/io/papermc/paper/registry/package-info.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..0b80179ff90e085568d7ceafd9b17511789dc99b
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/registry/package-info.java
+@@ -0,0 +1,5 @@
++@DefaultQualifier(NonNull.class)
++package io.papermc.paper.registry;
++
++import org.checkerframework.checker.nullness.qual.NonNull;
++import org.checkerframework.framework.qual.DefaultQualifier;
+diff --git a/src/main/java/net/minecraft/core/MappedRegistry.java b/src/main/java/net/minecraft/core/MappedRegistry.java
+index 6e6d9573115eeedef4cb24beac43f1bc96eebb49..befe662974bd3d6b4d0a2d69f9bfef46ec93986b 100644
+--- a/src/main/java/net/minecraft/core/MappedRegistry.java
++++ b/src/main/java/net/minecraft/core/MappedRegistry.java
+@@ -80,6 +80,14 @@ public class MappedRegistry<T> implements WritableRegistry<T> {
+             return MappedRegistry.this.getTags().map(Pair::getSecond);
+         }
+     };
++    // Paper start
++    @Nullable
++    private Map<ResourceLocation, T> beforeFrozen = new HashMap<>(2048);
```

Comment:
2000 is a weird default here, as a greta majority of registries are under 100, with the notable exception of items, blocks, sounds, and block states